### ISSUE TITLE
Canvas blur after rescale page

### ISSUE THREAD

**Issue by vitonsky** [fabricjs/fabric.js#7502]
Canvas is not rescale content after resize page.
When we load web page with 50% page zoom and then set zoom to 100% canvas looks  very bad like low resolution pucture.

I did try to manually rescale fabricjs inner objects while resize page like that:
```ts
export const fixCanvasSize = (canvas: fabric.Canvas) => {
    const lowerCanvasEl = (canvas as any).lowerCanvasEl;

    const width = (canvas as any).width;
    const height = (canvas as any).height;

    lowerCanvasEl.width = width;
    lowerCanvasEl.height = height;

    canvas.renderAll();
};
```

And this fix problem, but controls to resize objects are render wrong and i can't fix it.
If you have solution, tell about it.

Seems it impossible to fix on library user side

## Version
3.6.0

## Test Case
It reproduce on any page with fabricjs, try it on the http://fabricjs.com/

## Information about environment
All browsers

## Steps to reproduce
- Scale page with canvas to 50%
- Reload page
- Scale page to 100%

## Expected Behavior
- Fine canvas picture

## Actual Behavior
- Blur on the canvas, because it was render as 50% size and then resized without rescale

**Comment by rockerBOO** [comment]
Try using setDimensions

```
   canvas.setDimensions({
      width: window.innerWidth,
      height: window.innerHeight
    });
```

**Comment by vitonsky** [comment]
@rockerBOO i've try it, but it not work for me.

I did try use only `setDimensions`, also tried set size for `lowerCanvasEl` and then call `setDimensions` and vice versa.
I also tried all options combinations. It did not fix this bug.

After call `setDimensions` canvas resize without fix of scale factor.

**Comment by rockerBOO** [comment]
```
  canvas.setWidth(window.innerWidth);
  canvas.setHeight(window.innerHeight);

  canvas.calcOffset();
  canvas.requestRenderAll();
```

I have also used this option. Don't try to adjust the lowerCanvasEl or upperCanvasEl. They are generally abstracted away where you shouldn't need to access them. You can infer things from them but manually adjusting is probably going to cause some conflict.

**Comment by vitonsky** [comment]
@rockerBOO unfortunately `calcOffset` is not fix canvas too. Is it work for you?
Btw, i not resize canvas object, i only rescale web page (but i set size explicitly for test it as you say)

I adjust `lowerCanvasEl` because this method is fix a image, but after this fix an object controls render with incorrect position.
I think that fabric isn't rescale `lowerCanvasEl` itself and have not hook to do it manual.
Under rescale i mean not only change size, but change size such a way to consider resolution, display size and page scale.

I've tried `calcViewportBoundaries` to fix controls, but it not work too.

**Comment by vitonsky** [comment]
@rockerBOO is you just set new size while `resize` event and it work for you? Did you have high resolution image after this manipulation? And if you will not do it you have low resolution image on canvas after rescale page?

**Comment by rockerBOO** [comment]
Slapped this example together.

```html
<!DOCTYPE html>
<html>
<head>
<script src="https://unpkg.com/fabric@4.6.0/dist/fabric.min.js"></script>
<style>
body { margin: 0; }
</style>
</head>
	<body>
		<canvas id="canvas1"></canvas>
<script>
		window.onload = () => {
	const canvas = new fabric.Canvas("canvas1");

canvas.add(new fabric.Circle({ radius: 100, fill: 'red', top: 100, left: 100 }))
const onResize = () => {
		canvas.setWidth(window.innerWidth)
		canvas.setHeight(window.innerHeight)
}
window.addEventListener('resize', onResize)

		onResize()
}
</script>
	</body>
</html>
```

**Comment by vitonsky** [comment]
@rockerBOO i tried [your example](https://codesandbox.io/s/wizardly-dawn-olzqf?file=/index.html)

I have blured image:
![](https://user-images.githubusercontent.com/86191922/144496302-9e8fe15f-5e49-4626-b974-256992132042.png)

Reproduce steps:
- scale page to 50%
- reload page
- reset scale of page to 100%

**Comment by rockerBOO** [comment]
Ahh yes I see what you're saying now. Sorry about that. I have noticed this as long as I have used it but just thought it was a canvas thing.

**Comment by vitonsky** [comment]
@rockerBOO it's fine. What you think about it? It seems like bug?
What we can do to fix it?

**Comment by asturur** [comment]
hi!
The issue here is that the pagezoom is mixed with retina scaling.
to be honest retina scale values that are lesser than 1 should be ignored to avoid blurryness.

**Comment by vitonsky** [comment]
@asturur hi. Can we implement some knob to manual update values which affect on this problem while render?
Or maybe we can fix this bug automatically inside fabricjs, if it possible.

I think little things like this is very important for high quality projects.

Thanks for attention

**Comment by asturur** [comment]
we should avoid setting retina scaling to less than 1 in fabricjs, since there are no browsers that have more css pixels than real pixels.
The bug would persist anyway, if you are a page zoom 200, and then set back to 100% you would have a larger canvas.

i think you can open a pr by yourself, the function is called getRetinaScaling and should never return a number smaller than 1.

**Comment by vitonsky** [comment]
@asturur thanks for explaining bug mechanic.

I've create PR https://github.com/fabricjs/fabric.js/pull/7537 with fix

Also i want to leave here solution for people from future who have this bug.

As told @asturur cause in `devicePixelRatio` property of fabricjs which is not equal with actual `devicePixelRatio` of window, so you should manually update this value, reinitialize retina scaling and render all like that:

```ts
const onResize = () => {
  // update devicePixelRatio
  fabric.devicePixelRatio = fabric.window.devicePixelRatio ||
                          fabric.window.webkitDevicePixelRatio ||
                          fabric.window.mozDevicePixelRatio ||
                          1;
  canvas._initRetinaScaling();

  // re-render canvas with new quality
  canvas.requestRenderAll();
}

window.addEventListener('resize', onResize)
```

It's should fix problem. Also, fabric is provide a `setDimensions` method which run `_initRetinaScaling` and recalc other dynamic properties itself, so maybe it will better to call

**Comment by asturur** [comment]
on my projects i attach a resize event on the canvas container and then i use setDimension to adapt the canvas.

**Comment by vitonsky** [comment]
@asturur yes, call `setDimension` is better than inner method `_initRetinaScaling`.

Btw, what you think about some hook for `fabric` to update `devicePixelRatio` ?
Maybe something more abstract, like `reinitializeInnerState` which execute logic like in first initialization of fabric object.

It may be useful for cases like this bug, when some inner variable is more not actual after some changes which we can't predict.
We will just call something like `fabric.reinitializeInnerState()` and variables like `fabric.devicePixelRatio` will recalculate explicitly.

**Comment by ShaMan123** [comment]
Thanks guys, I think this solves an issue in my project.
+1 to update device pixel ratio on resize **internally**, [`_onResize`](https://github.com/fabricjs/fabric.js/blob/7efc732d8a050d743a66c89d8b1c4e5f942191f4/src/mixins/canvas_events.mixin.js#L372)
